presto-spi/src/main/java/com/facebook/presto/spi/block/MapBlockBuilder.java

```diff
@@ -0,0 +1,267 @@
+
+public class MapBlockBuilder
+        extends AbstractMapBlock
+        implements BlockBuilder
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(MapBlockBuilder.class).instanceSize() + BlockBuilderStatus.INSTANCE_SIZE;
+
+    private final MethodHandle keyBlockHashCode;
+
+    private BlockBuilderStatus blockBuilderStatus;
+
+    private int positionCount;
+    private int[] offsets;
+    private boolean[] valueIsNull;
+    private final BlockBuilder keyBlockBuilder;
+    private final BlockBuilder valueBlockBuilder;
+    private int[] hashTable;
+
+    private int currentEntrySize;
+
+    public MapBlockBuilder(
+            Type keyType,
+            Type valueType,
+            MethodHandle keyNativeHashCode,
+            MethodHandle keyBlockHashCode,
+            MethodHandle keyBlockNativeEquals,
+            BlockBuilderStatus blockBuilderStatus,
+            int expectedEntries)
+    {
+        super(keyType, keyNativeHashCode, keyBlockNativeEquals);
+
+        this.keyBlockHashCode = requireNonNull(keyBlockHashCode, "keyBlockHashCode is null");
+        this.blockBuilderStatus = requireNonNull(blockBuilderStatus, "blockBuilderStatus is null");
+
+        this.positionCount = 0;
+        this.offsets = new int[expectedEntries + 1];
+        this.valueIsNull = new boolean[expectedEntries];
+        this.keyBlockBuilder = keyType.createBlockBuilder(blockBuilderStatus, expectedEntries);
+        this.valueBlockBuilder = valueType.createBlockBuilder(blockBuilderStatus, expectedEntries);
+        this.hashTable = new int[16];
+        Arrays.fill(hashTable, -1);
+    }
+
+    @Override
+    protected Block getKeys()
+    {
+        return keyBlockBuilder;
+    }
+
+    @Override
+    protected Block getValues()
+    {
+        return valueBlockBuilder;
+    }
+
+    @Override
+    protected int[] getHashTable()
+    {
+        return hashTable;
+    }
+
+    @Override
+    protected int[] getOffsets()
+    {
+        return offsets;
+    }
+
+    @Override
+    protected int getOffsetBase()
+    {
+        return 0;
+    }
+
+    @Override
+    protected boolean[] getValueIsNull()
+    {
+        return valueIsNull;
+    }
+
+    @Override
+    public int getPositionCount()
+    {
+        return positionCount;
+    }
+
+    @Override
+    public int getSizeInBytes()
+    {
+        return keyBlockBuilder.getSizeInBytes() + valueBlockBuilder.getSizeInBytes() +
+                (Integer.BYTES + Byte.BYTES) * positionCount +
+                Integer.BYTES * HASH_MULTIPLIER * keyBlockBuilder.getPositionCount();
+    }
+
+    @Override
+    public int getRetainedSizeInBytes()
+    {
+        return intSaturatedCast(
+                INSTANCE_SIZE + keyBlockBuilder.getRetainedSizeInBytes() + valueBlockBuilder.getRetainedSizeInBytes() + sizeOf(offsets) + sizeOf(valueIsNull) + sizeOf(hashTable));
+    }
+
+    @Override
+    public MapElementBlockWriter beginBlockEntry()
+    {
+        if (currentEntrySize != 0) {
```

Comment:
It is true that `currentEntrySize` can only be 0 or 1. I'm ok with changing it to a `boolean`. However, I'm not sure `entryOpen` is a good name, given the ambiguity of open/closed in this context.